### PR TITLE
Integrating grafana go sdk

### PR DESCRIPTION
**Pull request by bmanth60**
Added integration from https://github.com/valeriakononenko/grafana-clickhouse-plugin/tree/master/

This is to solve: #242

### CODE REVIEW

File: docker-compose.yaml [hunk 1]
```diff
@@ -40,6 +40,7 @@ services:
       # @TODO sign instructions in https://grafana.com/docs/grafana/latest/developers/plugins/sign-a-plugin/
       # - GF_PLUGINS_ALLOW_LOADING_UNSIGNED_PLUGINS=vertamedia-clickhouse-datasource
       - GF_INSTALL_PLUGINS=grafana-piechart-panel,grafana-worldmap-panel
+      - GF_APP_MODE=development
```
Comment (bmanth60) [inline docker-compose.yaml:43]: Not sure if this should be enabled by default

File: docker-compose.yaml [hunk 2]
```diff
@@ -97,8 +99,7 @@ services:
     command: |
       bash -xec "
         if [[ ! -f ./node_modules/@grafana/toolkit/bin/grafana-toolkit.js ]]; then npm install --production=false; fi && \
-        source /grafana-clickhouse/.release_env && \
-        node ./node_modules/@grafana/toolkit/bin/grafana-toolkit.js plugin:sign
+        npm run sign
```
Comment (bmanth60) [inline docker-compose.yaml:102]: Removed duplicate maintenance for this command

File: backend/value.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+				return NewValue(&timeValue, fieldName, []*time.Time{})
+			}
+		}
+	}
+
+	return NullValue(fieldName, []*time.Time{})
+}
+
+func ParseValue(valueType string, value interface{}, fieldName string, timezone *time.Location) *Value {
```
Comment (bmanth60) [inline backend/value.go:90]: Need to add Tuple parsing

File: backend/main.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+	"encoding/json"
+	"errors"
+	"fmt"
+	"os"
+
+	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/datasource"
+)
+
+func parseJson(jsonData []byte, res interface{}) error {
```
Comment (bmanth60) [inline backend/main.go:13]: Not sure we should be wrapping json.Unmarshal

File: backend/field.go [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+	FrameField *data.Field
+	Name       string
+	Type       string
+	TimeZone   *time.Location
+}
+
+func (f *ClickHouseField) Append(value interface{}) {
+	v := ParseValue(f.Type, value, f.Name, f.TimeZone)
+	if v == nil {
+		backend.Logger.Warn(fmt.Sprintf("Value [%v / %v] wouln't be added to Field", value, f.Type))
```
Comment (bmanth60) [inline backend/field.go:59]: typo on wouldn't

File: backend/client.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+
+	req, err := http.NewRequest(
+		"POST",
+		dataSourceUrl,
+		bytes.NewBufferString(query))
+	if err != nil {
+		return onErr(err)
+	}
+
+	req.Header.Set("X-ClickHouse-User", client.settings.Username)
```
Comment (bmanth60) [inline backend/client.go:41]: This removed parsing for our other settings such as withcredentials/usePost vs use get etc. that needs to be added back in.

File: backend/client.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+		backend.Logger.Error(fmt.Sprintf("clickhouse client query error: %w", err)
+		return nil, err
+	}
+
+	datasourceUrl, err := url.Parse(client.context.URL)
+	if err != nil {
+		return onErr(fmt.Errorf("unable to parse clickhouse dataSourceUrl: %w", err))
+	}
+
+	httpClient := &http.Client{}
```
Comment (bmanth60) [inline backend/client.go:31]: this should abstract a function that creates our fully formed client

File: docker-compose.yaml (hunk 1 above)
Comment (Slach) [reply docker-compose.yaml:43]: which problem do you try to resolve to add `development` here?

File: docker-compose.yaml (hunk 2 above)
Comment (Slach) [reply docker-compose.yaml:102]: nice catch thanks

**Comment by Slach** [comment]
good work, we should add parsing for corner case with `Tuple(SeriesName String, value Numeric)`

**Comment by Slach** [comment]
@bmanth60 did you receive GRAFANA_API_KEY over email?

**Comment by bmanth60** [comment]
@Slach yes I did, thanks! I also see that I'm added to the org. Will sort this out asap.

File: docker-compose.yaml (hunk 1 above)
Comment (bmanth60) [reply docker-compose.yaml:43]: This was to load the plugin but it seems that the signing is working now so I do not need this flag. I have removed.

**Comment by Slach** [comment]
@bmanth60 any progress in this PR ?

**Comment by bmanth60** [comment]
@Slach not particularly. I have it working for a very basic case in our test alerts. It still doesn't work for array/tuples. Also, it doesn't read basic auth. There is still a lot of work to be done. I have been reading https://github.com/grafana/github-datasource and would be great if we could end up in a similar position. Right now the plugin is handlers are dispersed, so it should be reaggregated.

**Comment by Slach** [comment]
@bmanth60 hello =) glad to see your efforts 
I hope you can resolve conflicts with `master` branch, just accept your version of binaries and manifest.txt
will you support in your branch Array(String, float) which produced after $columns macro?

**Comment by bmanth60** [comment]
@Slach The intention is to have at least what is in master functional before attempting to merge this

**Comment by bmanth60** [comment]
@Slach I think I finally cracked this. Can you please review and test? I think it is working as expected. I won't be able to get unit tests in this PR, that will be for another journey. I have a fixture in my local if you want me to commit that.

**Comment by Slach** [comment]
@bmanth60 thanks a lot for your efforts, i will test it tomorrow
could you resolve conflicts from master?
```
git remote add upstream git@github.com:Vertamedia/clickhouse-grafana.git
git pull upstream master
```
and accept yours binaries version during conflict resolve?

**Comment by bmanth60** [comment]
done

**Comment by Slach** [comment]
@bmanth60 I tested your branch locally
look like `$macros` in queries works with little error
steps to reproduce

## New SDK
```bash
wget https://github.com/bmanth60/clickhouse-grafana/archive/refs/heads/convert.to.sdk.zip
unzip clickhouse-grafana-convert.to.sdk.zip
cd clickhouse-grafana-convert.to.sdk
docker-compose up -d clickhouse grafana
firefox http://localhost:3000/d/Rln7U1GGk/test-alerts-dashboard?editPanel=2&orgId=1&refresh=5s
docker-compose down
```
image

## Old SDK (2.2.3)
```bash
wget https://github.com/Vertamedia/clickhouse-grafana/archive/refs/tags/v2.2.3.zip
unzip clickhouse-grafana-2.2.3.zip
cd clickhouse-grafana-2.2.3
docker-compose up -d clickhouse grafana
firefox http://localhost:3000/d/Rln7U1GGk/test-alerts-dashboard?editPanel=2&orgId=1&refresh=5s
docker-compose down
```
image

File: pkg/frame.go [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+			field.Fields[0] = NewField(compoundName, compoundType, tz)
+		}
+
+		// Parse the last compound item
+		if field.IsCompound {
+			compoundMeta := strings.Split(fieldType, ",")
+			compoundParts := make([]*ClickHouseField, len(compoundMeta))
+			for i := 0; i < len(compoundMeta); i++ {
+				compoundType := strings.TrimSpace(compoundMeta[i])
+				compoundParts[i] = NewField(compoundName+"-"+compoundType, compoundType, tz)
```
Comment (Slach) [inline pkg/frame.go:49]: when we get compound type after apply $macros, like Array(Tupple()
string element of Tuple is a Time Series Name, numeric element of Tuple is a series value
please look to https://github.com/Vertamedia/clickhouse-grafana/blob/v2.2.3/datasource.go#L287-L323 for an explanation

**Review by Slach — CHANGES_REQUESTED** [review]
almost look great ;) thanks a lot for your effort and I personally really appreciated, only little bit changes require for parsing `Array(Tuple(` data types properly

**Comment by bmanth60** [comment]
@Slach I did some refactoring and added the parsing for series type where  `Array(Tuple(String` has the series name as string. Please rereview.

**Comment by Slach** [comment]
@bmanth60 tested successful with 6.7.6 and 7.3.10 and 7.5.1
thanks a lot for your effort

**Pull request merged**